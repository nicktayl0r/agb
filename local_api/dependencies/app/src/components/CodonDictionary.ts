export const CODON_DICTIONARY: { [index: string]: string } = {
  UUU: "Phenylalanine",
  UUC: "Phenylalanine",
  UUA: "Leucine",
  UUG: "Leucine",
  CUU: "Leucine",
  CUC: "Leucine",
  CUA: "Leucine",
  CUG: "Leucine",
  AUU: "Isoleucine",
  AUC: "Isoleucine",
  AUA: "Isoleucine",
  AUG: "Methionine",
  GUU: "Valine",
  GUC: "Valine",
  GUA: "Valine",
  GUG: "Valine",
  UCU: "Serine",
  UCC: "Serine",
  UCA: "Serine",
  UCG: "Serine",
  CCU: "Proline",
  CCC: "Proline",
  CCA: "Proline",
  CCG: "Proline",
  ACU: "Threonine",
  ACC: "Threonine",
  ACA: "Threonine",
  ACG: "Threonine",
  GCU: "Alanine",
  GCC: "Alanine",
  GCA: "Alanine",
  GCG: "Alanine",
  UAU: "Tyrosine",
  UAC: "Tyrosine",
  UAA: "Stop",
  UAG: "Stop",
  CAU: "Histidine",
  CAC: "Histidine",
  CAA: "Glutamine",
  CAG: "Glutamine",
  AAU: "Asparagine",
  AAC: "Asparagine",
  AAA: "Lysine",
  AAG: "Lysine",
  GAU: "Aspartic acid",
  GAC: "Aspartic acid",
  GAA: "Glutamic acid",
  GAG: "Glutamic acid",
  UGU: "Cysteine",
  UGC: "Cysteine",
  UGA: "Stop_UGA",
  UGG: "Tryptophan",
  CGU: "Arginine",
  CGC: "Arginine",
  CGA: "Arginine",
  CGG: "Arginine",
  AGU: "Serine",
  AGC: "Serine",
  AGA: "Arginine",
  AGG: "Arginine",
  GGU: "Glycine",
  GGC: "Glycine",
  GGA: "Glycine",
  GGG: "Glycine"
};
