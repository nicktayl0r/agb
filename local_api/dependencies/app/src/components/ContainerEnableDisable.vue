<template>
  <fieldset :id="id" v-show="visibleStore" :disabled="disabled" :class="[ 'no-border', AddRemoveClasses]" ><slot/></fieldset>
</template>

<script lang="ts">
import { Component } from "vue-property-decorator";
import Widget from "./Widget";
import WidgetEnabled from "../mixins/WidgetEnabled";

@Component({ mixins: [WidgetEnabled] })
export default class ContainerEnableDisable extends Widget {}
</script>
