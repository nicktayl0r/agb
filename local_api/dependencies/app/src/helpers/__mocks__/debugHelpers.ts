const mockDebugHelpers = require.requireActual("@/helpers/debugHelpers");

mockDebugHelpers.isDebug = () => true;
mockDebugHelpers.logInfo = (message: string, ...optionalParams: any[]) => { };
mockDebugHelpers.logMessage = (message: string, ...optionalParams: any[]) => { };
mockDebugHelpers.logWarning = (message: string, ...optionalParams: any[]) => { };
mockDebugHelpers.logError = (message: string, ...optionalParams: any[]) => { };
mockDebugHelpers.gizmoLogSecurely = (message: string, ...optionalParams: any[]) => { };
mockDebugHelpers.logWarningSecurely = (message: string, ...optionalParams: any[]) => { };
mockDebugHelpers.logErrorSecurely = (error: string, ...optionalParams: any[]) => { };

module.exports = mockDebugHelpers;
