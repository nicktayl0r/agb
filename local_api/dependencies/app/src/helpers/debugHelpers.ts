import * as Sentry from "@sentry/browser";

import { captureSentryEvent, captureSentryException } from "@/helpers/sentryHelper";

import { getEnv } from "@/helpers/envHelpers";
import { getStore } from "@/helpers/storeHelpers";
import { getPageID } from "@/helpers/widgetHelpers";
import { readModulesState } from "@/store/modules/userData";

import { readSessionId, readAttemptId } from "@/store/modules/authData";

import {
  readProjectName,
  readPlayerGitVersion,
  readProjectGitVersion
} from "@/store/modules/appData";

import { vueApp } from "..";

import { AxiosRetryInstance } from "@/helpers/requestHelpers";

import { canAccessLocalStorage } from "@/helpers/browserHelpers";

import { pageIsUnloading } from "@/store/mutationHistoryPlugin";

let debug = false;

export function setDebug(_debug: boolean) {
  debug = _debug;
}

export function isDebug() {
  return debug || window.location.href.includes("?debug=true");
}

export function showHandbookPages() {
  return window.location.href.includes("?handbookPages=true");
}

export function showDebugLabel() {
  return !window.location.href.includes("?debugLabel=false");
}

const filterValues = [
  "elClassID",
  "response",
  "responsedata",
  "comments",
  "score",
  "progress",
  "state",
  "attemptId",
  "totalscore",
  "sessionId",
  "userActivityAttemptId",
  "userActivityResponseId"
];


function combineOptionalParams(optionalParams: any[]) {
  let returnString = "";

  if (optionalParams != undefined) {
    for (const param of optionalParams) {
      if (param != undefined) returnString += ` ${stringifyParam(param)}`;
    }
  }

  return returnString;
}

function stringifyParam(param: any) {
  if (typeof param === "object" && param !== undefined) {
    return `\n${JSON.stringify(param, undefined, 4)}\n`;
  }
  return param.toString();
}

// clean up objects for the browser log, w/o Vue getters and setters
function removeFunctionsFromObjects(params: any[]) {
  const newArray: any[] = [];
  for (const index in params) {
    if (typeof params[index] === "object" && params[index] !== undefined) {
      newArray[index] = JSON.parse(JSON.stringify(params[index]));
    } else newArray[index] = params[index];
  }
  return newArray;
}

export function logMessage(message: string, ...optionalParams: any[]) {
  if (isDebug() || getEnv() === "development") {
    console.log(message, ...removeFunctionsFromObjects(optionalParams));
    // const stringifiedMessage = message + combineOptionalParams(optionalParams);
    // console.log(stringifiedMessage);
  }
}

export function logInfo(_message: string, ...optionalParams: any[]) {
  console.log(_message, ...removeFunctionsFromObjects(optionalParams));

  if (!pageIsUnloading()) {
    const moduleState = readModulesState(getStore());
    captureSentryEvent({
      message: _message,
      level: Sentry.Severity.Info,
      tags: {
        session_trackGroup: moduleState.currentTrackGroupID,
        session_page: getPageID(vueApp),
        local_storage_enabled: canAccessLocalStorage().toString()
      },
      extra: { ...removeFunctionsFromObjects(optionalParams) }
    });
  }
}

export function logError(_message: string, ...optionalParams: any[]) {
  console.error(_message, ...removeFunctionsFromObjects(optionalParams));

  if (!pageIsUnloading()) {
    const moduleState = readModulesState(getStore());
    captureSentryEvent({
      message: _message,
      level: Sentry.Severity.Error,
      tags: {
        session_trackGroup: moduleState.currentTrackGroupID,
        session_page: getPageID(vueApp),
        local_storage_enabled: canAccessLocalStorage().toString()
      },
      extra: { ...removeFunctionsFromObjects(optionalParams) }
    });
  }
}

export function logWarning(_message: string, ...optionalParams: any[]) {
  console.warn(_message, ...removeFunctionsFromObjects(optionalParams));

  if (!pageIsUnloading()) {
    const moduleState = readModulesState(getStore());
    captureSentryEvent({
      message: _message,
      level: Sentry.Severity.Warning,
      tags: {
        session_trackGroup: moduleState.currentTrackGroupID,
        session_page: getPageID(vueApp),
        local_storage_enabled: canAccessLocalStorage().toString()
      },
      extra: { ...removeFunctionsFromObjects(optionalParams) }
    });
  }
}

export function logException(error: any, ...optionalParams: any[]) {
  console.error(error, ...removeFunctionsFromObjects(optionalParams));

  if (!pageIsUnloading()) {
    const stringifiedError = error;
    captureSentryException(stringifiedError);
  }
}

export function gizmoLogSecurely(message: string, ...optionalParams: any[]) {
  console.log(message);
  sendGizmoLogForm("INFO", message, optionalParams);
}

export function logWarningSecurely(message: string, ...optionalParams: any[]) {
  console.warn(message);

  if (!pageIsUnloading()) {
    sendGizmoLogForm("WARN", message, optionalParams);

    const moduleState = readModulesState(getStore());
    captureSentryEvent({
      message: `[GizmoLog]${message}`,
      level: Sentry.Severity.Warning,
      tags: {
        session_trackGroup: moduleState.currentTrackGroupID,
        session_page: getPageID(vueApp),
        local_storage_enabled: canAccessLocalStorage().toString()
      },
      extra: { ...removeFunctionsFromObjects(optionalParams) }
    });
  }
}

export function logExceptionSecurely(error: any, ...optionalParams: any[]) {
  console.error(error);

  if (!pageIsUnloading()) {
    const stringifiedError = error;
    sendGizmoLogForm("ERROR", stringifiedError, optionalParams);
    captureSentryException(new Error(`[GizmoLog]${stringifiedError}`));
  }
}

export function logErrorSecurely(message: string, ...optionalParams: any[]) {
  console.error(message);
  if (!pageIsUnloading()) {
    sendGizmoLogForm("ERROR", message, optionalParams);

    const moduleState = readModulesState(getStore());
    captureSentryEvent({
      message: `[GizmoLog]${message}`,
      level: Sentry.Severity.Error,
      tags: {
        session_trackGroup: moduleState.currentTrackGroupID,
        session_page: getPageID(vueApp),
        local_storage_enabled: canAccessLocalStorage().toString()
      },
      extra: { ...removeFunctionsFromObjects(optionalParams) }
    });
  }
}

async function sendGizmoLogForm(
  messageType: string,
  message: string,
  ...optionalParams: any[]
) {
  //send form request
  // URL: https://api.explorelearning.com/api/v3/write-to-log
  // Staging  URL: https://test-api.explorelearning.com/api/v3/write-to-log

  let url = undefined;
  switch (getEnv()) {
    case "production":
      url = "https://api.explorelearning.com/api/v3/write-to-log";
      break;
    case "stage":
      url = "https://test-api.explorelearning.com/api/v3/write-to-log";
      break;
    // case "development":
    //   url= "http://dev-api.explorelearning.com/api/v3/write-to-log";
    //   break;
    // case "debug":
    //   //url= "http://dev-api.explorelearning.com/api/v3/write-to-log";
    //   url= "";
    //   break;
    default:
      url = "https://api.explorelearning.com/api/v3/write-to-log";
      break;
  }

  const bodyFormData = new FormData();

  try {
    bodyFormData.append("MessageType", messageType);
    bodyFormData.append(
      "Message",
      `${message} ${combineOptionalParams(optionalParams)}
      AttemptID: ${readAttemptId(getStore())} SessionID: ${readSessionId(getStore())}
      Project: ${readProjectName(getStore())}
      ProjectGitVersion: ${readProjectGitVersion(
        getStore()
      )} PlayerGitVersion: ${readPlayerGitVersion(getStore())}`
    );
    bodyFormData.append("Key", "SoylentGreen1974");
    bodyFormData.append("Subsystem", `STEMCase_VueApp_${getEnv()}`);
  } catch (error) {
    logError(`loadSimState error: ${error}`);
  }

  const headers = {
    config: { headers: { "Content-Type": "multipart/form-data" } }
  };

  if (url) {
    AxiosRetryInstance
      .post(url, bodyFormData)
      .catch((errors): any => {
        console.error("Cannot Use Gizmo Logs", errors);
        // commented below out because it was hogging sentry with junk errors
        // logError("Cannot Use Gizmo Logs", errors, headers);
      });
  }
}
