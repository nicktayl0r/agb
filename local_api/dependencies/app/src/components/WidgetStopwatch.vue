<template>
	<span :id="id" v-show="visibleStore" :class="AddRemoveClasses" :start_ms="start_msStore" :end_ms="end_msStore" :play-state="playStateStore" :speed="speedStore" :autoplay="autoPlayStore">
		{{ formattedTimeString }}
	</span>
</template>

<script lang="ts">
import { Component, Prop } from "vue-property-decorator";

import { logMessage, logWarning } from "@/helpers/debugHelpers";
import { runWidgetEventConditions } from "@/helpers/widgetHelpers";
import WidgetPlayable, { state } from "@/mixins/WidgetPlayable";
import { readWidgetEntryVal } from "@/store/modules/userData";

import Widget from "./Widget";
import { setInterval, clearInterval } from "timers";

@Component({ mixins: [WidgetPlayable] })
export default class WidgetStopwatch extends Widget {
	@Prop({ default: 0 }) start_ms: number; //milliseconds

	@Prop({ default: 10000 }) end_ms: number; //milliseconds

	@Prop({ default: 1 }) speed: number;

	@Prop({ default: false }) autoPlay: boolean;

	@Prop({ default: "colon" }) format: "colon" | "hrs/mins" | "mins" | "secs";

	@Prop() sharedDataWriteKey: string;

	currentElapsedMS = 0;

	pastElapsedMS = 0; // use this to store time elapsed in previous plays, that may have been at different speeds

	// these are in js Date format: the number of milliseconds elapsed since January 1, 1970 00:00:00 UTC.
	startJSDate = 0;

	nowJSDate = 0;

	// window.startInterval and window.clearInterval use this number as an ID to reference the interval function
	intervalHandler: number | null = null;

	mounted() {
	  // @ts-ignore playStateStore is in WidgetPlayable
	  if (this.playStateStore != "play" && this.autoPlayStore) this.updateWidget("playState", "play");

	  this.pastElapsedMS = 0;
	}

	get start_msStore() {
	  return this.propStoreOverride("start_ms") as number;
	}

	get end_msStore() {
	  return this.propStoreOverride("end_ms") as number;
	}

	get autoPlayStore() {
	  return this.propStoreOverride("autoPlay") as boolean;
	}

	prevSpeed = 1;

	get speedStore() {
	  const newSpeed = this.propStoreOverride("speed") as number;
	  if (this.prevSpeed !== newSpeed) this.currentToPastElapsed();
	  this.prevSpeed = newSpeed;
	  return newSpeed;
	}

	get formattedTimeString(): string {
	  switch (this.$props.format) {
	  case "hrs/mins":
	    return (
	      `${(this.totalMS < 0 ? "-" : "")
					+ this.hours
					 } hrs ${
					 this.minutes
					 } mins`
	    );
	  case "colon":
	    let val = "";
	    if (this.totalMS < 0) val += "-";
	    if (this.hours != "00") val += `${this.hours}:`;
	    if (this.minutes != "00") val += `${this.minutes}:`;
	    val += `${this.seconds}:${this.hundredthSeconds}`;
	    return val;
	  case "mins":
	    return this.totalminutes;
	  case "secs":
	    return this.totalseconds;
	  }
	  return "";
	}

	get totalElapsedMS() {
	  return this.pastElapsedMS + this.currentElapsedMS;
	}

	get hours(): string {
	  const trim = this.totalMS >= 0 ? Math.floor : Math.ceil;

	  // const hrs = Math.abs(
	  // 	this.countDown
	  // 		? trim((this.start_msStore - this.totalElapsedMS) / 1000 / 60 / 60)
	  // 		: trim(this.totalElapsedMS / 1000 / 60 / 60)
	  // );
	  const hrs = Math.abs(trim(this.totalMS / 1000 / 60 / 60));
	  return hrs >= 10 ? hrs.toString() : `0${hrs}`;
	}

	get minutes(): string {
	  const trim = this.totalMS >= 0 ? Math.floor : Math.ceil;
	  // const min = Math.abs(
	  // 	this.countDown
	  // 		? trim(((this.start_msStore - this.totalElapsedMS) / 1000 / 60) % 60)
	  // 		: trim((this.totalElapsedMS / 1000 / 60) % 60)
	  // );
	  const min = Math.abs(trim((this.totalMS / 1000 / 60) % 60));
	  return min >= 10 ? min.toString() : `0${min}`;
	}

	get seconds(): string {
	  const trim = this.totalMS >= 0 ? Math.floor : Math.ceil;
	  // const sec = Math.abs(
	  // 	this.countDown
	  // 		? trim(((this.start_msStore - this.totalElapsedMS) / 1000) % 60)
	  // 		: trim((this.totalElapsedMS / 1000) % 60)
	  // );
	  const sec = Math.abs(trim((this.totalMS / 1000) % 60));
	  return sec >= 10 ? sec.toString() : `0${sec}`;
	}

	get hundredthSeconds(): string {
	  const trim = this.totalElapsedMS >= 0 ? Math.ceil : Math.floor;
	  // const hund = Math.abs(
	  // 	this.countDown
	  // 		? trim(((this.start_msStore - this.totalElapsedMS) / 10) % 100)
	  // 		: trim((this.totalElapsedMS / 10) % 100)
	  // );
	  const hund = Math.abs(trim((this.totalMS / 10) % 100));

	  if (hund == 100) return "00";
	  return hund >= 10 ? hund.toString() : `0${hund}`;
	}

	get totalminutes(): string {
	  // const min = this.countDown
	  // 	? Math.floor((this.start_msStore - this.totalElapsedMS) / 1000 / 60)
	  // 	: Math.floor(this.totalElapsedMS / 1000 / 60);
	  const min = Math.floor(this.totalMS / 1000 / 60);
	  return min.toString();
	}

	get totalseconds(): string {
	  // const secs = this.countDown
	  // 	? Math.floor((this.start_msStore - this.totalElapsedMS) / 1000)
	  // 	: Math.floor(this.totalElapsedMS / 1000);

	  const secs = Math.floor(this.totalMS / 1000);
	  return secs.toString();
	}

	get totalMS() {
	  return this.countDown
	    ? this.maxMS - this.totalElapsedMS
	    : this.totalElapsedMS + this.minMS;
	}

	get durationMS() {
	  return this.maxMS - this.minMS;
	}

	get countDown() {
	  return this.start_msStore >= this.end_msStore;
	}

	get minMS() {
	  if (!this.countDown) return this.start_msStore;
	  return this.end_msStore;
	}

	get maxMS() {
	  if (!this.countDown) return this.end_msStore;
	  return this.start_msStore;
	}

	DoPlayState(s: state) {
	  this.stopInterval();
	  logMessage("stopwatch DoPlayState", s);

	  if (this.totalElapsedMS >= this.durationMS) return; // prevent effects causing a play/end infinite loop

	  this.startJSDate = Date.now();
	  this.nowJSDate = this.startJSDate;
	  this.currentElapsedMS = 0;
	  const prevElapsed = readWidgetEntryVal(this.$store)(
	    this.pageID,
	    this.$props.id,
	    "elapsedMS"
	  );
	  logMessage("prevElapsed", prevElapsed.toString());
	  if (prevElapsed.isJust()) this.pastElapsedMS = prevElapsed.unsafelyUnwrap() as number;

	  logMessage("currentElapsedMS", this.currentElapsedMS.toString());
	  logMessage("totalElapsedMS", this.totalElapsedMS.toString());

	  this.StoreTime();

	  this.intervalHandler = window.setInterval(this.updateCurrentTime, 10); // 10ms is .01 seconds
	}

	DoStopState(s: state) {
	  this.stopInterval();

	  this.startJSDate = Date.now();
	  this.nowJSDate = this.startJSDate;
	  this.currentElapsedMS = 0;

	  this.StoreTime();
	}

	DoPauseState(s: state) {
	  this.stopInterval();

	  this.StoreTime();
	  this.currentToPastElapsed();
	}

	currentToPastElapsed() {
	  const currentVal = this.currentElapsedMS;
	  this.currentElapsedMS = 0;
	  this.pastElapsedMS += currentVal;
	  this.startJSDate = Date.now();
	}

	DoEnd() {
	  this.stopInterval();

	  this.StoreTime();
	  runWidgetEventConditions(this, this.$props.endEffects, this.$props.id);
	}

	StoreTime() {
	  if (this.$props.sharedDataWriteKey) {
	    this.updateSharedDataVal(
	      this.$props.sharedDataWriteKey,
	      this.formattedTimeString
	    );
	  }
	}

	beforeDestroy() {
	  this.StoreTime();
	}

	updateCurrentTime() {
	  // @ts-ignore playStateStore is in WidgetPlayable
	  //console.log("updateCurrentTime");
	  if (this.playStateStore === "play") {
	    this.nowJSDate = Date.now();
	    this.currentElapsedMS =				(this.nowJSDate - this.startJSDate) * this.speedStore;

	    // console.log(
	    // 	`this.nowJSDate ${this.nowJSDate} this.currentElapsedMS ${this.currentElapsedMS}`
	    // );
	    if (this.totalElapsedMS >= this.durationMS) {
	      //logMessage("totalElapsedMS", this.totalElapsedMS, "durationMS", this.durationMS);
	      logMessage("totalMS", this.totalMS, "maxMS", this.maxMS);
	      this.currentElapsedMS = this.durationMS + this.pastElapsedMS; // don't go over duration
	      // @ts-ignore currentPlayState is in WidgetPlayable
	      this.currentPlayState = "default"; // necessary to get ApplyPlayState to run play events after this
	      this.updateWidget("playState", "default");
	      this.DoEnd();
	    }
	  }
	}

	stopInterval() {
	  if (this.intervalHandler) window.clearInterval(this.intervalHandler);
	  this.intervalHandler = null;
	}
}
</script>