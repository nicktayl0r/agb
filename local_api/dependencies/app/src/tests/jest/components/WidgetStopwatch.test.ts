import { getLocalVue, getCleanStore } from "@/tests/jest/jestSetup";
const localVue = getLocalVue();
let store = getCleanStore();

import { mount, shallowMount } from "@vue/test-utils";
import { setDebug } from "@/helpers/debugHelpers.ts";
import { WidgetBaseTests } from "@/tests/jest/components/WidgetBaseTests";
import { MixinWidgetPlayableTests } from "@/tests/jest/mixins/MixinWidgetPlayableTests";

import WidgetStopwatch from "@/components/WidgetStopwatch.vue";
import {
  readWidgetEntryVal,
  dispatchUpdateWidget
} from "@/store/modules/userData";

describe("WidgetStopwatch", () => {
  WidgetBaseTests(WidgetStopwatch);
  MixinWidgetPlayableTests(WidgetStopwatch);
});

describe("WidgetStopwatch snapshot", () => {
  beforeEach(() => {
    store = getCleanStore();
  });

  it.each`
		_format        | _start     | _end       | _setDateNowTo
		${"secs"}     | ${0}      | ${5000}   | ${4000}
		${"secs"}     | ${5000}   | ${0}      | ${4000}
		${"mins"}     | ${0}      | ${300000} | ${240000}
		${"mins"}     | ${300000} | ${0}      | ${240000}
		${"colon"}    | ${0}      | ${5000}   | ${4125}
		${"colon"}    | ${5000}   | ${0}      | ${3725}
		${"colon"}    | ${0}      | ${300000} | ${240125}
		${"colon"}    | ${300000} | ${0}      | ${239275}
		${"hrs/mins"} | ${0}      | ${300000} | ${240000}
		${"hrs/mins"} | ${300000} | ${0}      | ${240000}
	`("stopwatch snapshots", ({
    _format, _start, _end, _setDateNowTo
  }) => {
    const wrapper = shallowMount(WidgetStopwatch, {
      propsData: {
        id: "testWidgetStopwatch",
        text: "stopwatch text",
        start_ms: _start,
        end_ms: _end,
        speed: 1,
        autoPlay: false,
        format: _format,
        playState: "default"
      },
      localVue,
      store
    });

    const { vm } = wrapper;
    const { updateCurrentTime } = (vm as any);

    dispatchUpdateWidget(store, {
      pageID: "testPage",
      widgetID: "testWidgetStopwatch",
      key: "playState",
      value: "play"
    });

    updateCurrentTime();
    global.Date.now = jest.fn(() => _setDateNowTo);
    updateCurrentTime();

    expect(wrapper.element).toMatchSnapshot();
  });
  //add more snapshots for each distinctive property configuration.
});

describe("WidgetStopwatch functionality", () => {
  beforeEach(() => {
    store = getCleanStore();

    global.Date.now = jest.fn(() => 0);
    setDebug(true);

    dispatchUpdateWidget(store, {
      pageID: "testPage",
      widgetID: "testWidgetStopwatch",
      key: "elapsedMS",
      value: 0
    });
  });

  it.each`
		start   | end
		${0}    | ${5000}
		${5000} | ${0}
	`("Goto playState state", ({ start, end }) => {
    const wrapper = shallowMount(WidgetStopwatch, {
      propsData: {
        id: "testWidgetStopwatch",
        start_ms: start,
        end_ms: end,
        speed: 1,
        autoPlay: false,
        format: "secs",
        playState: "default"
      },
      localVue,
      store
    });

    dispatchUpdateWidget(store, {
      pageID: "testPage",
      widgetID: "testWidgetStopwatch",
      key: "playState",
      value: "play"
    });

    const widgetVal = readWidgetEntryVal(store)(
      "testPage",
      "testWidgetStopwatch",
      "playState"
    );
    expect(widgetVal.unsafelyUnwrap()).toBe("play");
  });

  it.each`
		start   | end     | countDown
		${0}    | ${5000} | ${false}
		${5000} | ${0}    | ${true}
	`("CountDown = start > end?", ({ start, end, countDown }) => {
    const wrapper = shallowMount(WidgetStopwatch, {
      propsData: {
        id: "testWidgetStopwatch",
        start_ms: start,
        end_ms: end,
        speed: 1,
        autoPlay: false,
        format: "secs",
        playState: "default"
      },
      localVue,
      store
    });
    const { vm } = wrapper;
    expect((vm as any).countDown).toBe(countDown);
  });

  //todo: add a speed parameter to fully represent the combinations of properties
  it.each`
		_format        | _start      | _end       | _setDateNowTo  | _formattedTimeString  | _totalMS   | _speed
		${"secs"}     | ${1000}    | ${5000}   | ${3000}       | ${"4"}               | ${4000}   | ${1}
		${"secs"}     | ${5000}    | ${1000}   | ${3000}       | ${"2"}               | ${2000}   | ${1}
		${"secs"}     | ${-1000}   | ${5000}   | ${3000}       | ${"2"}               | ${2000}   | ${1}
		${"secs"}     | ${-5000}   | ${1000}   | ${3000}       | ${"-2"}              | ${-2000}  | ${1}
		${"secs"}     | ${300000}  | ${1000}   | ${240000}     | ${"60"}              | ${60000}  | ${1}
		${"mins"}     | ${1000}    | ${300000} | ${240000}     | ${"4"}               | ${241000} | ${1}
		${"mins"}     | ${300000}  | ${1000}   | ${240000}     | ${"1"}               | ${60000}  | ${1}
		${"mins"}     | ${-1000}   | ${300000} | ${240000}     | ${"3"}               | ${239000} | ${1}
		${"mins"}     | ${-300000} | ${1000}   | ${240000}     | ${"-1"}              | ${-60000} | ${1}
		${"colon"}    | ${1000}    | ${5000}   | ${2125}       | ${"03:13"}           | ${3125}   | ${1}
		${"colon"}    | ${5000}    | ${1000}   | ${3725}       | ${"01:28"}           | ${1275}   | ${1}
		${"colon"}    | ${-1000}   | ${5000}   | ${2125}       | ${"01:13"}           | ${1125}   | ${1}
		${"colon"}    | ${-5000}   | ${1000}   | ${3725}       | ${"-01:27"}          | ${-1275}  | ${1}
		${"colon"}    | ${1000}    | ${300000} | ${240125}     | ${"04:01:13"}        | ${241125} | ${1}
		${"colon"}    | ${300000}  | ${1000}   | ${238275}     | ${"01:01:73"}        | ${61725}  | ${1}
		${"hrs/mins"} | ${1000}    | ${300000} | ${239000}     | ${"00 hrs 04 mins"}  | ${240000} | ${1}
		${"hrs/mins"} | ${300000}  | ${1000}   | ${240000}     | ${"00 hrs 01 mins"}  | ${60000}  | ${1}
		${"hrs/mins"} | ${-1000}   | ${300000} | ${239000}     | ${"00 hrs 03 mins"}  | ${238000} | ${1}
		${"hrs/mins"} | ${-300000} | ${1000}   | ${240000}     | ${"-00 hrs 01 mins"} | ${-60000} | ${1}
		${"secs"}     | ${1000}    | ${5000}   | ${1500}       | ${"4"}               | ${4000}   | ${2}
		${"secs"}     | ${5000}    | ${1000}   | ${1500}       | ${"2"}               | ${2000}   | ${2}
		${"secs"}     | ${-1000}   | ${5000}   | ${1500}       | ${"2"}               | ${2000}   | ${2}
		${"secs"}     | ${-5000}   | ${1000}   | ${1500}       | ${"-2"}              | ${-2000}  | ${2}
		${"secs"}     | ${300000}  | ${1000}   | ${120000}     | ${"60"}              | ${60000}  | ${2}
		${"mins"}     | ${1000}    | ${300000} | ${120000}     | ${"4"}               | ${241000} | ${2}
		${"mins"}     | ${300000}  | ${1000}   | ${120000}     | ${"1"}               | ${60000}  | ${2}
		${"mins"}     | ${-1000}   | ${300000} | ${120000}     | ${"3"}               | ${239000} | ${2}
		${"mins"}     | ${-300000} | ${1000}   | ${120000}     | ${"-1"}              | ${-60000} | ${2}
		${"colon"}    | ${1000}    | ${5000}   | ${1062}       | ${"03:13"}           | ${3125}   | ${2}
		${"colon"}    | ${5000}    | ${1000}   | ${3725 / 2}   | ${"01:28"}           | ${1275}   | ${2}
		${"colon"}    | ${-1000}   | ${5000}   | ${1062}       | ${"01:13"}           | ${1125}   | ${2}
		${"colon"}    | ${-5000}   | ${1000}   | ${3725 / 2}   | ${"-01:27"}          | ${-1275}  | ${2}
		${"colon"}    | ${1000}    | ${300000} | ${240125 / 2} | ${"04:01:13"}        | ${241125} | ${2}
		${"colon"}    | ${300000}  | ${1000}   | ${238275 / 2} | ${"01:01:73"}        | ${61725}  | ${2}
		${"hrs/mins"} | ${1000}    | ${300000} | ${239000 / 2} | ${"00 hrs 04 mins"}  | ${240000} | ${2}
		${"hrs/mins"} | ${300000}  | ${1000}   | ${240000 / 2} | ${"00 hrs 01 mins"}  | ${60000}  | ${2}
		${"hrs/mins"} | ${-1000}   | ${300000} | ${239000 / 2} | ${"00 hrs 03 mins"}  | ${238000} | ${2}
		${"hrs/mins"} | ${-300000} | ${1000}   | ${240000 / 2} | ${"-00 hrs 01 mins"} | ${-60000} | ${2}
	`(
    "formattedTimeString - Format=$format; $start -> $end after $setDateNowTo MS",
    ({
      _format,
      _start,
      _end,
      _setDateNowTo,
      _formattedTimeString,
      _totalMS,
      _speed
    }) => {
      const wrapper = shallowMount(WidgetStopwatch, {
        propsData: {
          id: "testWidgetStopwatch",
          start_ms: _start,
          end_ms: _end,
          speed: _speed,
          autoPlay: false,
          format: _format,
          playState: "default"
        },
        localVue,
        store
      });
      const { vm } = wrapper;
      const { updateCurrentTime } = (vm as any);

      dispatchUpdateWidget(store, {
        pageID: "testPage",
        widgetID: "testWidgetStopwatch",
        key: "playState",
        value: "play"
      });

      updateCurrentTime();
      global.Date.now = jest.fn(() => _setDateNowTo);
      updateCurrentTime();

      // expect(vm.totalElapsedMS).toBe(setDateNowTo);
      // expect(vm.totalMS).toBe(totalMS);
      expect((vm as any).formattedTimeString).toBe(_formattedTimeString);
    }
  );
});
