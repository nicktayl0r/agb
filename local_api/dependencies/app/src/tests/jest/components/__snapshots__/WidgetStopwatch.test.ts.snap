// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WidgetStopwatch snapshot stopwatch snapshots 1`] = `
<span
  class=""
  end_ms="5000"
  id="testWidgetStopwatch"
  play-state="default"
  speed="1"
  start_ms="0"
  text="stopwatch text"
>
  
	0

</span>
`;

exports[`WidgetStopwatch snapshot stopwatch snapshots 2`] = `
<span
  class=""
  end_ms="0"
  id="testWidgetStopwatch"
  play-state="default"
  speed="1"
  start_ms="5000"
  text="stopwatch text"
>
  
	5

</span>
`;

exports[`WidgetStopwatch snapshot stopwatch snapshots 3`] = `
<span
  class=""
  end_ms="300000"
  id="testWidgetStopwatch"
  play-state="default"
  speed="1"
  start_ms="0"
  text="stopwatch text"
>
  
	0

</span>
`;

exports[`WidgetStopwatch snapshot stopwatch snapshots 4`] = `
<span
  class=""
  end_ms="0"
  id="testWidgetStopwatch"
  play-state="default"
  speed="1"
  start_ms="300000"
  text="stopwatch text"
>
  
	5

</span>
`;

exports[`WidgetStopwatch snapshot stopwatch snapshots 5`] = `
<span
  class=""
  end_ms="5000"
  id="testWidgetStopwatch"
  play-state="default"
  speed="1"
  start_ms="0"
  text="stopwatch text"
>
  
	00:00

</span>
`;

exports[`WidgetStopwatch snapshot stopwatch snapshots 6`] = `
<span
  class=""
  end_ms="0"
  id="testWidgetStopwatch"
  play-state="default"
  speed="1"
  start_ms="5000"
  text="stopwatch text"
>
  
	05:00

</span>
`;

exports[`WidgetStopwatch snapshot stopwatch snapshots 7`] = `
<span
  class=""
  end_ms="300000"
  id="testWidgetStopwatch"
  play-state="default"
  speed="1"
  start_ms="0"
  text="stopwatch text"
>
  
	00:00

</span>
`;

exports[`WidgetStopwatch snapshot stopwatch snapshots 8`] = `
<span
  class=""
  end_ms="0"
  id="testWidgetStopwatch"
  play-state="default"
  speed="1"
  start_ms="300000"
  text="stopwatch text"
>
  
	05:00:00

</span>
`;

exports[`WidgetStopwatch snapshot stopwatch snapshots 9`] = `
<span
  class=""
  end_ms="300000"
  id="testWidgetStopwatch"
  play-state="default"
  speed="1"
  start_ms="0"
  text="stopwatch text"
>
  
	00 hrs 00 mins

</span>
`;

exports[`WidgetStopwatch snapshot stopwatch snapshots 10`] = `
<span
  class=""
  end_ms="0"
  id="testWidgetStopwatch"
  play-state="default"
  speed="1"
  start_ms="300000"
  text="stopwatch text"
>
  
	00 hrs 05 mins

</span>
`;
